### PR TITLE
version 0.9.0

### PR DESCRIPTION
**Pull request by Khoroshevskyi**
version 0.9.0 is ready for release.

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -24,11 +25,16 @@ You can add `--processed` if you want to download processed files from the given
 geofetch -i GSE95654 --processed
 ```
 
+
 You can add `--just-metadata` if you want to download metadata without the raw SRA files or processed GEO files.
 
 ```console
 geofetch -i GSE95654 --just-metadata
 ```
```
Comment (nleroy917) [inline docs/README.md:34]: Should we add something here like: 

"You can also combine the `--processed` and `--just-metadata` flags:"

File: geofetch/geofetch.py [hunk 2]
```diff
@@ -782,6 +757,29 @@ def unify_list_keys(self, processed_meta_list):
...
+        genome_keys = [
+            "Assembly",
+            "assembly",
+            "ASSEMBLY",
+            "Genome_build",
+            "genome_build",
+            "genome build",
+            "Genome build",
+        ]
+        proj_gen_keys = list(set(list_keys).intersection(genome_keys))
```
Comment (nleroy917) [inline geofetch/geofetch.py:774]: Instead of having every permutation of casing in the above `genome_keys` list, could we just coerce the `list_keys` to lowercase and replace all white_space with an underscore?

```python
proj_gen_keys = list(set(
[k.lower().strip().replace(" ", "_") for k in list_keys]
).intersection(genome_keys))
```

Then `genome_keys` becomes:

```
[
    "assembly",
    "genome_build",
]

File: geofetch/geofetch.py [hunk 3]
```diff
@@ -1590,15 +1800,25 @@ def _parse_cmdl(cmdl):
...
         default=None,
-        help="Optional: Specify one or more filepaths to pipeline interface yaml files. "
+        help="Optional: Specify one or more filepaths to SAMPLES pipeline interface yaml files. "
         "These will be added to the project config file to make it immediately "
         "compatible with looper. [Default: null]",
     )
 
+    # Optional
+    parser.add_argument(
+        "--pipeline_project",
```
Comment (nleroy917) [inline geofetch/geofetch.py:1814]: should shis be `--pipeline-project`?

File: geofetch/geofetch.py [hunk 4]
```diff
@@ -1590,15 +1800,25 @@ def _parse_cmdl(cmdl):
         "--config-template", default=None, help="Project config yaml file template."
     )
 
+    # Optional
     parser.add_argument(
-        "-P",
-        "--pipeline_interfaces",
+        "--pipeline_samples",
```
Comment (nleroy917) [inline geofetch/geofetch.py:1805]: Should this be `--pipeline-samples`?

File: geofetch/utils.py [hunk 5]
```diff
@@ -293,3 +298,21 @@ def convert_size(size_str: str) -> int:
         size_in_bytes = int(size_numb[0]) * abbreviation_dict[abb]
```
Comment (nleroy917) [inline geofetch/utils.py:298]: Do we need a case here for a sizing unit that **isnt** included in the dictionary? Something like:

```python
s_unit=abbreviation_dict.get(abb)

if s_unit is not None:
    # continue on
else:
    # default to something or raise useful exception
```

**Review by nleroy917 — COMMENTED** [review]
I think some of the new `cli` flags are `--in_this_form` and need to be `--in-this-form`. I also had some small thoughts on other things to try and handle any weird edge cases.

**Comment by nleroy917** [comment]
A feature I think would be cool/useful for pephub is to be able to do at least one of the following:

1. specify the name of the PEP config file, or
2. create a `.pep.yaml` file which points to the config file in the folder

I see that there is a flag for name:

```
-n NAME, --name NAME  Specify a project name
```

But I think that just specifies the project name in the `.yaml` file. It'd be cool (helpful) to be able to specify the name of the file itself. PEPhub currently looks for `project_config.yaml`, and if that doesn't exist it looks for a `.pep.yaml` file that tells it where the config file is.

**Comment by nleroy917** [comment]
It looks like the `.soft` files were not removed after processing, even though I used the `--discard-soft` flag:

```
find out/ -name "*.soft" | wc -l
> 572
```

 Here is an example of a `.soft` file that is quite large:

image

The sample table also doesn't appear truncated. Inspecting `GSE143256_samples.csv` inside excel shows:

image

Which is 510 characters. Here is what I ran:

```
geofetch -i accessions.txt \
--processed \
--just-metadata \
--discard-soft \
--attr-limit-truncate 100
```

I might be running the `cli` wrong, though. Let me know! I love that it now makes a separate folder for each PEP and also creates a `.pep.yaml` file, that is super useful.

**Comment by Khoroshevskyi** [comment]
> It looks like the `.soft` files were not removed after processing, even though I used the `--discard-soft` flag:
> 
> ```
> find out/ -name "*.soft" | wc -l
> > 572
> ```
> 
> Here is an example of a `.soft` file that is quite large:
> 
> image
> 
> The sample table also doesn't appear truncated. Inspecting `GSE143256_samples.csv` inside excel shows:
> 
> image
> 
> Which is 510 characters. Here is what I ran:
> 
> ```
> geofetch -i accessions.txt \
> --processed \
> --just-metadata \
> --discard-soft \
> --attr-limit-truncate 100
> ```
> 
> I might be running the `cli` wrong, though. Let me know! I love that it now makes a separate folder for each PEP and also creates a `.pep.yaml` file, that is super useful.

I have no idea why you have bug there. I checked it one more time, and everything works correctly/:

File: geofetch/geofetch.py (hunk 3 above)
Comment (Khoroshevskyi) [reply geofetch/geofetch.py:1814]: @nsheff,  I've devided --pipleine-interface into two arguments: `--pipeline-project` and `--pipeline-samples`. does it sounds good?

Comment (Khoroshevskyi) [reply geofetch/geofetch.py:1814]: > should shis be `--pipeline-project`?

Yes it should be --pipeline-project, thx, I haven't noticed it

File: geofetch/geofetch.py (hunk 4 above)
Comment (Khoroshevskyi) [reply geofetch/geofetch.py]: Corrected

File: geofetch/geofetch.py (hunk 3 above)
Comment (Khoroshevskyi) [reply geofetch/geofetch.py]: Corrected

File: geofetch/utils.py (hunk 5 above)
Comment (Khoroshevskyi) [reply geofetch/utils.py:298]: Added Exception

File: geofetch/geofetch.py (hunk 2 above)
Comment (Khoroshevskyi) [reply geofetch/geofetch.py:782]: Fixed!:)

**Comment by nleroy917** [comment]
> Why do you have date May 17 ??

That is a great question, I can try to run it again and see if I get a similar result.

**Comment by nsheff** [comment]
Maybe it's an older file, and so there's a different issue -- it's not that it's not truncating or doing what it says, it's failing earlier and just not overwriting your old file. could the soft files be from a previous run?

**Comment by Khoroshevskyi** [comment]
> It was an older file I believe. It's running again now. Is there a flag to create a separate folder for each PEP? It l looks like it finished by just creating one big POP.

image

**Comment by nleroy917** [comment]
This is small, but I noticed all the files have sample tales with attributes that start with a capital letter. Is it possible to make them fully lowercase?

e.g. "Sample_name" -> "sample_name"

**Comment by Khoroshevskyi** [comment]
> This is small, but I noticed all the files have sample tales with attributes that start with a capital letter. Is it possible to make them fully lowercase?
> 
> e.g. "Sample_name" -> "sample_name"

For processed metadata this issue is solved already. For raw metadata it won't be solved in this version.

**Comment by nleroy917** [comment]
> > This is small, but I noticed all the files have sample tales with attributes that start with a capital letter. Is it possible to make them fully lowercase?
> > e.g. "Sample_name" -> "sample_name"
> 
> For processed metadata this issue is solved already. For raw metadata it is not planed to be solved in this version.

Ok!

**Comment by Khoroshevskyi** [comment]
> Adding the `--acc-anon` flag pushes each PEP into an individual folder. So that l looks to be working.
> 
> > It'd be cool (helpful) to be able to specify the name of the file itself. PEPhub currently looks for project_config.yaml, and if that doesn't exist it looks for a .pep.yaml file that tells it where the config file is.
> 
> Is this possible? This would make it **super** easy to integrate with pephub. Since, pephub requires that a PEP folder has that `project_config.yaml` file or a `.pep.yaml` file pointing to the right one.
> 
> @nsheff should pephub just try to load the first `.yaml` file it sees?

It's possible, but I am not sure if we should to do this, as in one folder we can have few PEPs.
What we can do is: change second part of the project name. 
Right now we have: 
image
it's for processed data

**Comment by nsheff** [comment]
I see value in a CLI argument to specify the *project config* file name.

However, what will this do if you use `--acc-anno`, which is going to produce multiple PEPs?

**Comment by nleroy917** [comment]
## Meeting June 14th, 2022 Notes
Two things to accomplish:
1. Ability to create `.pep.yaml` in the folder where PEPs are stored
2. Series data can no longer be stored in the same folder

For the creation of the `.pep.yaml` file, maybe a flag name can be: `--add-dotfile`

**Comment by nleroy917** [comment]
Works great so far with the `--add-dotfile` flag!

**Comment by Khoroshevskyi** [comment]
@nsheff I think now geofetch is ready for new release. Additionally, we can add some infromation to documentation including fugures, do you have any suggestions about it?

File: docs/changelog.md [hunk 6]
```diff
@@ -1,5 +1,18 @@
 # Changelog
 
+## [0.9.0] -- 2022-06-20
+- Updated New pipeline interface.
+- Fixed error sample_name error when title does not exist.
```
Comment (nsheff) [inline docs/changelog.md:5]: Can you rephrase this? It doesn't make sense.

File: docs/changelog.md [hunk 7]
```diff
@@ -1,5 +1,18 @@
...
+- Added `--discard-soft` argument.
+- Added `--const-limit-project` argument.
+- Added `--const-limit-discard` argument.
+- Added `--attr-limit-truncate` argument.
+- Added `"--add-dotfile"` argument.
+- Disabled creating combined pep when flag `--acc-anno` is set.
+- Improved finding and separating metadata keys and genotype information
```
Comment (nsheff) [inline docs/changelog.md:12]: genotype information? what exactly is geofetch doing with genotype? Do you mean genome assembly?

File: geofetch/geofetch.py [hunk 8]
```diff
@@ -131,18 +135,28 @@ def __init__(
         self.refresh_metadata = refresh_metadata
         self.config_template = config_template
 
+        # if user specified a pipeline interface path for samples, add it into the project config
+        if pipeline_samples and pipeline_samples != "null":
+            self.file_pipeline_samples = pipeline_samples
+            self.file_pipeline_samples = (
+                "pipeline_interfaces: ;l" + self.file_pipeline_samples
```
Comment (nsheff) [inline geofetch/geofetch.py:142]: what is meant by the `;1` here? I'm not familiar with this

**Comment by nsheff** [comment]
The processed data tutorial has an abnormal list rendering: 

image

probably missing a newline after a list title, or something

File: geofetch/geofetch.py (hunk 8 above)
Comment (Khoroshevskyi) [reply geofetch/geofetch.py]: I think I typed this by mistake. Fixed.

File: docs/changelog.md (hunk 7 above)
Comment (Khoroshevskyi) [reply docs/changelog.md]: Fixed

File: docs/changelog.md (hunk 6 above)
Comment (Khoroshevskyi) [reply docs/changelog.md]: fixed

File: docs/changelog.md [hunk 9]
```diff
@@ -1,5 +1,18 @@
 # Changelog
 
+## [0.9.0] -- 2022-06-20
+- Updated New pipeline interface.
```
[the remark this answers is not in the document]
Comment (Khoroshevskyi) [reply docs/changelog.md]: Done

**Comment by nsheff** [comment]
@nleroy917 are you OK with this release?

**Review by nsheff — APPROVED** [review]

**Comment by nleroy917** [comment]
Yes. Looks good to me!
